src/lib.rs

```diff
@@ -49,6 +49,19 @@ pub enum NumberOrString {
     String(String),
 }
 
+/* ----------------- Message typing ----------------- */
+
+pub trait Request {
+    type Params;
+    type Result;
+    const METHOD: &'static str;
+}
+
+pub trait Notification {
+    type Params;
+    const METHOD: &'static str;
+}
```

Comment:
I wonder, would it be better to move the traits and their implementations to submodules so they are not confused with the data holding types? `mod rpc`, `mod notification + mod request`?

Reply:
Yeah, actually the thought crossed my mind as well. I'll try to separate it into `mod notification + mod request`.

Reply:
@Marwes should I move there helper union return values that aren't defined by the protocol, but needed to encode the result type? e.g. https://github.com/gluon-lang/languageserver-types/pull/29/files#diff-b4aea3e418ccdb71239b96952d9cddb6R2000

Reply:
I could see either place being good. Maybe an argument for keeping them here is that it makes the submodules only hold types that you aren't meant to instantiate (since they don't hold data).

Reply:
Keeping them here means the result type you linked to.